### ISSUE TITLE
Checksums on instance values

### ISSUE THREAD

**Issue by jmakeig** [marklogic/entity-services#176]
As a compliance officer, I want to be able to quickly and accurately determine if an entity instance has changed. Rather than doing a property-by-property comparison, it would be convenient to be able to hash entities to get a unique signature that reflects its complete state. If that state changes the hash must change too. It’s important that this hash is *not* dependent on the actual serialization of the entity  (i.e. whitespace, namespace prefixes, etc.), like `Object.hashCode()` in Java.

**Comment by grechaw** [comment]
You are right -- an in-memory entity instance could be accurately hashed.  Of course
that's not a general XML solution.  Entity services
could easily provide this functionality.   If properties appropriate for hashing were part of
the model declarations, then the hash could be used to deduplicate data too --
in your staging-to-production process just get unique values for the hash and only promote
that many objects.  Also, denormalized and standalone entity instances might share a hash
value...  nice idea

**Comment by grechaw** [comment]
If a denormalized instance's hash doesn't match the standalone copy, then a cascading update needs to be applied (or there's some other data problem)

**Comment by jmakeig** [comment]
> If a denormalized instance’s hash doesn't match the standalone copy, then a cascading update needs to be applied (or there's some other data problem)

@grechaw I’m not sure what you’re getting at here.

**Comment by grechaw** [comment]
Let's  say that you have a 'master' document for a person, but that person also exists, as denormalized copies, in some other documents for speed.  If you find that their two hashes are not the same, then the denormalized copies should be updated from the master record (in a corb-like job, say)

**Comment by jmakeig** [comment]
Yes, we should have enough information in the model to do batch (or even trigger-based) updates along the graph of related entities to ensure denormalized copies are up-to-date. Testing checksums would be a quick way to determine if an entity’s denormalized state needs to be updated in a foreign entity.
